core/src/main/java/se/tre/freki/query/QueryStringTranslator.java

```diff
@@ -53,19 +74,110 @@ public void exitQuery(@NotNull final se.tre.freki.query.SelectParser.QueryContex
     } catch (InterruptedException e) {
       throw new QueryException("Interrupted while waiting for metric", e);
     }
+    List<TimeSeriesIdPredicate> predicateList;
+    try {
+      predicateList = allAsList(futurePredicateList).get();
+    } catch (InterruptedException e) {
+      throw new QueryException("Interrupted while waiting for tags", e);
+    } catch (ExecutionException e) {
+      throw new QueryException("Error while fetching tags", e);
+    }
+
+    for (int i = 0; i < predicateList.size() / 2; i++) {
```

Comment:
Use an iterator and call `#next` twice instead of working with indexes like this. Using an iterator is more defensive since it will iterate over all elements and it will throw an exception if anything is missing.